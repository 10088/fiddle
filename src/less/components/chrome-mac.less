.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 23px;
    font-size: 1.4rem;

    small,
    p,
    span {
      user-select: none;
    }
  }
}
