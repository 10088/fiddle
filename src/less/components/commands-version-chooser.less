button.version-chooser.bp3-button {
  border-radius: 3px;
}
