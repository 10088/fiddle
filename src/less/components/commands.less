@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-1;
  -webkit-app-region: drag;
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding: 10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      button {
        margin-left: 5px;
      }
    }
  }

  .address-bar {
    height: 28px;
    width: 360px;
    transition: width 0.4s;
    display: flex;

    input,
    span {
      font-family: @fonts-common;
      font-weight: 300;
      font-size: 11px;
      cursor: auto;
      -webkit-app-region: no-drag;
    }

    input {
      flex-grow: 1;
      background: white;
      padding-left: 1rem;
      height: 2.6rem;
    }

    input:invalid {
      box-shadow: inset 0 0 0 1px #e4252d;
    }

    span {
      background: white;
      color: #1d1d1d;
      border-bottom-color: rgb(209, 209, 209);
      border-bottom-style: solid;
      border-bottom-left-radius: 4px;
      border-bottom-width: 1px;
      border-image-outset: 0;
      border-image-repeat: stretch;
      border-image-slice: 100%;
      border-image-source: none;
      border-image-width: 1;
      border-left-color: rgb(209, 209, 209);
      border-left-style: solid;
      border-left-width: 1px;
      border-top-color: rgb(209, 209, 209);
      border-top-left-radius: 4px;
      border-top-style: solid;
      border-top-width: 1px;
      display: inline-block;
      height: 1.2rem;
      letter-spacing: normal;
      margin: 0;
      padding: 6px 0 10px 10px;
      user-select: text;
      width: 132px;
      text-rendering: auto;
      text-shadow: none;
      text-transform: none;
    }

    &.empty {
      width: 200px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }
}
