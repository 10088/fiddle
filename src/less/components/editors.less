.mosaic-tile {
  animation: fadein 0.15s;
}

.mosaic-window-toolbar  {
  & > div {
    display: flex;
    justify-content: space-between;
    width: 100%;
    padding: 0 3px 0 29px;

    .mosaic-controls {
      transform: scale(0.7);
    }

    h5 {
      margin: 4px 0 0 0;
    }

    button {
      margin-left: 8px;
    }
  }
}

@keyframes fadein {
  from { opacity: 0; }
  to { opacity: 1; }
}
