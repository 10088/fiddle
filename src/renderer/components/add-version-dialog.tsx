import { observer } from 'mobx-react';
import * as path from 'path';
import * as React from 'react';
import * as semver from 'semver';

import { NpmVersion } from '../../interfaces';
import { getElectronNameForPlatform } from '../../utils/electron-name';
import { AppState } from '../state';
import { Dialog } from './dialog';

export interface AddVersionDialogProps {
  appState: AppState;
}

export interface AddVersionDialogState {
  isValidElectron: boolean;
  isValidVersion: boolean;
  file?: File;
  version: string;
}

/**
 * The "add version" dialog allows users to add custom builds of Electron.
 *
 * @class AddVersionDialog
 * @extends {React.Component<AddVersionDialogProps, AddVersionDialogState>}
 */
@observer
export class AddVersionDialog extends React.Component<AddVersionDialogProps, AddVersionDialogState> {
  constructor(props: AddVersionDialogProps) {
    super(props);

    this.state = {
      isValidVersion: false,
      isValidElectron: false,
      version: ''
    };

    this.onSubmit = this.onSubmit.bind(this);
    this.onClose = this.onClose.bind(this);
    this.onChangeFile = this.onChangeFile.bind(this);
    this.onChangeVersion = this.onChangeVersion.bind(this);
  }

  /**
   * Handles a change of the file input
   *
   * @param {React.ChangeEvent<HTMLInputElement>} event
   */
  public async onChangeFile(event: React.ChangeEvent<HTMLInputElement>) {
    const { files } = event.target;
    const { binaryManager } = this.props.appState;
    const file = files && files[0] ? files[0] : undefined;

    const isValidElectron = !!(file
      && file.path
      && await binaryManager.getIsDownloaded('custom', file.path));

    this.setState({ file, isValidElectron });
  }

  public onChangeVersion(event: React.ChangeEvent<HTMLInputElement>) {
    const version = event.target.value || '';
    const isValidVersion = !!semver.valid(version);

    this.setState({
      version,
      isValidVersion,
     });
  }

  /**
   * Handles the submission of the dialog
   *
   * @returns {Promise<void>}
   */
  public async onSubmit(): Promise<void> {
    const { file, version } = this.state;

    if (!file) return;

    const name = file.path
      .slice(-20)
      .split(path.sep)
      .slice(1)
      .join(path.sep);

    const toAdd: NpmVersion = {
      localPath: file.path,
      version,
      name
    };

    this.props.appState.addLocalVersion(toAdd);
    this.onClose();
  }

  /**
   * Closes the dialog
   */
  public onClose() {
    this.props.appState.isAddVersionDialogShowing = false;
    this.reset();
  }

  /**
   * Reset this component's state
   */
  public reset(): void {
    this.setState({
      isValidElectron: false,
      isValidVersion: false,
      version: '',
      file: undefined
    });
  }

  get buttons() {
    const canSubmit = this.state.isValidElectron && this.state.isValidVersion;

    return [
      (
        <button
          className='button'
          key='submit'
          disabled={!canSubmit}
          onClick={this.onSubmit}
        >
          Add
        </button>
      ), (
        <button
          className='cancel'
          key='cancel'
          onClick={this.onClose}
        >
          Cancel
        </button>
      )
    ];
  }

  public render() {
    const { isAddVersionDialogShowing } = this.props.appState;
    const dirOptions = { webkitdirectory: 'true' };

    return (
      <Dialog
        isShowing={isAddVersionDialogShowing}
        isShowingBackdrop={true}
        buttons={this.buttons}
        onClose={this.onClose}
        isCentered={true}
        className='add-version-dialog'
        key='add-version-dialog'
      >
        <label
          htmlFor='custom-electron-version'
          className='force-button'
        >
          Select the folder containing {getElectronNameForPlatform()}
        </label>
        <input
          type='file'
          onChange={this.onChangeFile}
          id='custom-electron-version'
          name='custom-electron-version'
          {...dirOptions}
        />
        {this.renderPath()}
      </Dialog>
    );
  }

  private renderPath(): JSX.Element | null {
    const { file, isValidElectron } = this.state;

    if (!file || !file.path) return null;

    const info = isValidElectron
      ? `We found an ${getElectronNameForPlatform()} in this folder.`
      : `We did not find a ${getElectronNameForPlatform()} in this folder...`;

    return (
      <>
        <input
          readOnly={true}
          value={file.path}
        />
        <p>
          {info}
        </p>
        {this.renderVersionInput()}
      </>
    );
  }

  private renderVersionInput(): JSX.Element | null {
    const { isValidElectron, isValidVersion, version } = this.state;

    if (!isValidElectron) return null;

    return (
      <>
        <p>
          Please specify a version, used for typings and the name.
          Must be <code>semver</code> compliant.
        </p>
        <input
          className={isValidVersion ? '' : 'hasError'}
          value={version}
          onChange={this.onChangeVersion}
          placeholder='4.0.0'
        />
      </>
    );
  }

}
