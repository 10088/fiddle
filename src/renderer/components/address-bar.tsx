import * as React from 'react';
import { observer } from 'mobx-react';
import * as Octokit from '@octokit/rest';
import * as classNames from 'classnames';

import { AppState } from '../state';
import { INDEX_HTML_NAME, MAIN_JS_NAME, RENDERER_JS_NAME } from '../constants';
import { idFromUrl } from '../../utils/gist';
import { reaction } from 'mobx';

export interface AddressBarProps {
  appState: AppState;
}

export interface AddressBarState {
  value: string;
}

@observer
export class AddressBar extends React.Component<AddressBarProps, AddressBarState> {
  constructor(props: AddressBarProps) {
    super(props);

    this.loadFiddle = this.loadFiddle.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);

    this.state = {
      value: this.props.appState.gistId
    };
  }

  /**
   * Handle the form's submit event, trying to load whatever
   * URL was entered.
   *
   * @param {React.SyntheticEvent<HTMLFormElement>} event
   * @memberof AddressBar
   */
  public async handleSubmit(event: React.SyntheticEvent<HTMLFormElement>) {
    event.preventDefault();

    this.props.appState.gistId = idFromUrl(this.state.value) || this.state.value;
    if (this.state.value) {
      this.loadFiddle();
    }
  }

  /**
   * Once the component mounts, we'll subscribe to gistId changes
   */
  public componentDidMount() {
    reaction(
      () => this.props.appState.gistId,
      (gistId: string) => this.setState({ value: gistId })
    );
  }

  /**
   * Handle the change event, which usually just updates the address bar's value
   *
   * @param {React.ChangeEvent<HTMLInputElement>} event
   * @memberof AddressBar
   */
  public handleChange(event: React.ChangeEvent<HTMLInputElement>) {
    this.setState({ value: idFromUrl(event.target.value) || event.target.value });
  }

  /**
   * Load a fiddle
   *
   * @returns
   * @memberof AddressBar
   */
  public async loadFiddle() {
    const { gistId } = this.props.appState;

    if (!confirm('Are you sure you want to load a new fiddle, all current progress will be lost?')) return;

    try {
      const octo = new Octokit();
      const gist = await octo.gists.get({
        gist_id: gistId,
        id: gistId,
      });

      window.ElectronFiddle.app.setValues({
        html: gist.data.files[INDEX_HTML_NAME].content,
        main: gist.data.files[MAIN_JS_NAME].content,
        renderer: gist.data.files[RENDERER_JS_NAME].content,
      });
    } catch (error) {
      console.warn(`Loading fiddle failed`, error);
    }
  }

  public render() {
    const { isUnsaved } = this.props.appState;
    const { value } = this.state;
    const className = classNames('address-bar', isUnsaved, { empty: !value });

    return (
      <form className={className} onSubmit={this.handleSubmit}>
        <span>https://gist.github.com/</span>
        <input
          key='addressbar'
          type='text'
          placeholder='...'
          value={value}
          onChange={this.handleChange}
        />
      </form>
    );
  }
}
