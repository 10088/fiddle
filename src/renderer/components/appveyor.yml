environment:
  matrix:
    - nodejs_version: "8"

init:
  - git config --global core.symlinks true

install:
  - ps: Install-Product node $env:nodejs_version x64
  - node --version
  - npm i

cache:
  - '%APPDATA%\npm-cache -> appveyor.yml'

test_script:
  - npm run lint
  - npm run test

build: off
