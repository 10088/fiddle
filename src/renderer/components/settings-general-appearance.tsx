import { shell } from 'electron';
import * as fsType from 'fs-extra';
import { observer } from 'mobx-react';
import * as path from 'path';
import * as React from 'react';

import { CONFIG_PATH } from '../../constants';
import { fancyImport } from '../../utils/import';
import { AppState } from '../state';
import { getAvailableThemes, THEMES_PATH } from '../themes';
import { LoadedFiddleTheme } from '../themes-defaults';

export interface AppearanceSettingsProps {
  appState: AppState;
}

export interface AppearanceSettingsState {
  themes: Array<LoadedFiddleTheme>;
}

/**
 * Settings content to manage GitHub-related preferences.
 *
 * @class GitHubSettings
 * @extends {React.Component<AppearanceSettingsProps, {}>}
 */
@observer
export class AppearanceSettings extends React.Component<
  AppearanceSettingsProps, AppearanceSettingsState
> {
  public constructor(props: AppearanceSettingsProps) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
    this.openThemeFolder = this.openThemeFolder.bind(this);

    this.state = {
      themes: []
    };

    getAvailableThemes().then((themes) => {
      this.setState({ themes });
    });
  }

  /**
   * Handle change, which usually means that we'd like update
   * the current theme.
   *
   * @param {React.ChangeEvent<HTMLSelectElement>} event
   */
  public handleChange(event: React.ChangeEvent<HTMLSelectElement>) {
    this.props.appState.setTheme(event.target.value);
  }

  public async openThemeFolder(): Promise<boolean> {
    const fs = await fancyImport<typeof fsType>('fs-extra');

    try {
      await fs.ensureDir(THEMES_PATH);
      await shell.showItemInFolder(THEMES_PATH);
      return true;
    } catch (error) {
      console.warn(`Appearance Settings: Could not open themes folder`);
      return false;
    }
  }

  /**
   * Render the theme options.
   *
   * @returns {Array<JSX.Element>}
   */
  public renderOptions(): Array<JSX.Element> {
    const { themes } = this.state;

    return themes.map(({ name, file }) => {
      return (
        <option value={file} key={file}>
          {name}
        </option>
      );
    });
  }

  public render() {
    const themePath = path.join(CONFIG_PATH, 'themes');

    return (
      <div>
        <h4>Appearance</h4>
        <label key='theme-label'>
          To add themes, add JSON theme files to <a
            id='open-theme-folder'
            onClick={() => this.openThemeFolder()}
          >
            <code>{themePath}</code>
          </a>.
        </label>
        <select
          className='select-themes'
          value={`${this.props.appState.theme}`}
          onChange={this.handleChange}
        >
          {this.renderOptions()}
        </select>
      </div>
    );
  }
}
