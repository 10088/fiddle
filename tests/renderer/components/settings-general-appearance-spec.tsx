import { shallow } from 'enzyme';
import * as React from 'react';
const { shell } = require('electron');

import { AppearanceSettings } from '../../../src/renderer/components/settings-general-appearance';

jest.mock('fs-extra');
jest.mock('../../../src/utils/import', () => ({
  fancyImport: async (p: string) => require(p)
}));

describe('AppearanceSettings component', () => {
  beforeEach(() => {
    this.store = {};
  });

  it('renders', () => {
    const wrapper = shallow(
      <AppearanceSettings appState={this.store} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('attempts to open the folder on click', () => {
    const wrapper = shallow(
      <AppearanceSettings appState={this.store} />
    );
    const instance: AppearanceSettings = wrapper.instance() as any;
    instance.openThemeFolder = jest.fn();

    wrapper.find('#open-theme-folder').simulate('click');
    expect(instance.openThemeFolder).toHaveBeenCalled();
  });

  it('attempts to open the folder on openThemeFolder', async () => {
    const wrapper = shallow(
      <AppearanceSettings appState={this.store} />
    );
    const instance: AppearanceSettings = wrapper.instance() as any;
    await instance.openThemeFolder();

    expect(shell.showItemInFolder).toHaveBeenCalled();
  });

  it('handles an error in openThemeFolder()', async () => {
    const wrapper = shallow(
      <AppearanceSettings appState={this.store} />
    );
    const instance: AppearanceSettings = wrapper.instance() as any;
    (shell as any).showItemInFolder.mockImplementationOnce(() => {
      throw new Error('Bwap');
    });

    expect(await instance.openThemeFolder()).toBe(false);
  });
});
