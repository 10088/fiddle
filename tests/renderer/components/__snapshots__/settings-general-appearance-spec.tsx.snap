// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppearanceSettings component renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AppearanceSettings
    appState={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h4>
          Appearance
        </h4>,
        <label>
          To add themes, add JSON theme files to 
          <a
            id="open-theme-folder"
            onClick={[Function]}
          >
            <code>
              ~/fake-user/.electron-fiddle/themes
            </code>
          </a>
          .
        </label>,
        <select
          className="select-themes"
          onChange={[Function]}
          value="undefined"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Appearance",
        },
        "ref": null,
        "rendered": "Appearance",
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": "theme-label",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "To add themes, add JSON theme files to ",
            <a
              id="open-theme-folder"
              onClick={[Function]}
            >
              <code>
                ~/fake-user/.electron-fiddle/themes
              </code>
            </a>,
            ".",
          ],
        },
        "ref": null,
        "rendered": Array [
          "To add themes, add JSON theme files to ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <code>
                ~/fake-user/.electron-fiddle/themes
              </code>,
              "id": "open-theme-folder",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "~/fake-user/.electron-fiddle/themes",
              },
              "ref": null,
              "rendered": "~/fake-user/.electron-fiddle/themes",
              "type": "code",
            },
            "type": "a",
          },
          ".",
        ],
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "select-themes",
          "onChange": [Function],
          "value": "undefined",
        },
        "ref": null,
        "rendered": Array [],
        "type": "select",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h4>
            Appearance
          </h4>,
          <label>
            To add themes, add JSON theme files to 
            <a
              id="open-theme-folder"
              onClick={[Function]}
            >
              <code>
                ~/fake-user/.electron-fiddle/themes
              </code>
            </a>
            .
          </label>,
          <select
            className="select-themes"
            onChange={[Function]}
            value="undefined"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Appearance",
          },
          "ref": null,
          "rendered": "Appearance",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": "theme-label",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "To add themes, add JSON theme files to ",
              <a
                id="open-theme-folder"
                onClick={[Function]}
              >
                <code>
                  ~/fake-user/.electron-fiddle/themes
                </code>
              </a>,
              ".",
            ],
          },
          "ref": null,
          "rendered": Array [
            "To add themes, add JSON theme files to ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <code>
                  ~/fake-user/.electron-fiddle/themes
                </code>,
                "id": "open-theme-folder",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "~/fake-user/.electron-fiddle/themes",
                },
                "ref": null,
                "rendered": "~/fake-user/.electron-fiddle/themes",
                "type": "code",
              },
              "type": "a",
            },
            ".",
          ],
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "select-themes",
            "onChange": [Function],
            "value": "undefined",
          },
          "ref": null,
          "rendered": Array [],
          "type": "select",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
